helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -220,6 +222,103 @@ protected static void setWorkflowContext(HelixManager manager, String workflowRe
   }
 
   /**
+   * Intialize the user content store znode setup
+   * @param propertyStore       zookeeper property store
+   * @param workflowJobResource the name of workflow or job
+   * @param record              the initial data
+   */
+  protected static void createUserContent(HelixPropertyStore propertyStore, String workflowJobResource,
+      ZNRecord record) {
+    propertyStore.create(Joiner.on("/")
+        .join(TaskConstants.REBALANCER_CONTEXT_ROOT, workflowJobResource,
+            TaskUtil.USER_CONTENT_NODE), record, AccessOption.PERSISTENT);
+  }
+
+  /**
+   * Get user defined workflow or job level key-value pair data
+   *
+   * @param manager             a connection to Helix
+   * @param workflowJobResource the name of workflow
+   * @param key                 the key of key-value pair
+   *
+   * @return null if there is no such pair, otherwise return a String
+   */
+  protected static String getWorkflowJobUserContent(HelixManager manager,
+      String workflowJobResource, String key) {
+    ZNRecord r = manager.getHelixPropertyStore().get(Joiner.on("/")
+            .join(TaskConstants.REBALANCER_CONTEXT_ROOT, workflowJobResource, USER_CONTENT_NODE), null,
+        AccessOption.PERSISTENT);
+    return r != null ? r.getSimpleField(key) : null;
+  }
+
+  /**
+   * Add an user defined key-value pair data to workflow or job level
+   *
+   * @param manager             a connection to Helix
+   * @param workflowJobResource the name of workflow or job
+   * @param key                 the key of key-value pair
+   * @param value               the value of key-value pair
+   */
+  protected static void addWorkflowJobUserContent(final HelixManager manager,
+      String workflowJobResource, final String key, final String value) {
+    final String path = Joiner.on("/")
+        .join(TaskConstants.REBALANCER_CONTEXT_ROOT, workflowJobResource, USER_CONTENT_NODE);
+
+    manager.getHelixPropertyStore().update(path, new DataUpdater<ZNRecord>() {
+      @Override public ZNRecord update(ZNRecord znRecord) {
+        ZNRecord r = manager.getHelixPropertyStore().get(path, null, AccessOption.PERSISTENT);
```

Comment:
you do not have to read znrecord in your updater again, the znRecord passed to you is supposed to be the lastest record.

What you do in updater is something like:

znRecord.setSimpleField(key, value);
return znRecord;

The property store will read the lastest znode, call the updater with this znode content, and write back with the new record returned from your updater.